File: app/src/main/java/com/theone/music/ui/fragment/user/LoginRegisterItemFragment.java
package com.theone.music.ui.fragment.user;//  ┏┓　　　┏┓
//┏┛┻━━━┛┻┓
//┃　　　　　　　┃
//┃　　　━　　　┃
//┃　┳┛　┗┳　┃
//┃　　　　　　　┃
//┃　　　┻　　　┃
//┃　　　　　　　┃
//┗━┓　　　┏━┛
//    ┃　　　┃                  神兽保佑
//    ┃　　　┃                  永无BUG！
//    ┃　　　┗━━━┓
//    ┃　　　　　　　┣┓
//    ┃　　　　　　　┏┛
//    ┗┓┓┏━┳┓┏┛
//      ┃┫┫　┃┫┫
//      ┗┻┛　┗┻┛

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;

import com.theone.common.constant.BundleConstant;
import com.theone.music.app.util.CacheUtil;
import com.theone.music.data.model.User;
import com.theone.music.databinding.FragmentLoginRegisterBinding;
import com.theone.music.ui.fragment.BaseFragment;
import com.theone.music.viewmodel.LoginRegisterViewModel;
import com.theone.mvvm.ext.qmui.QMUIFragmentExtKt;

/**
 * @author The one
 * @date 2022-04-11 14:36
 * @describe TODO
 * @email 625805189@qq.com
 * @remark
 */
public class LoginRegisterItemFragment extends BaseFragment<LoginRegisterViewModel, FragmentLoginRegisterBinding> {

    private LoginRegisterItemFragment(){}

    public static LoginRegisterItemFragment newInstance(boolean isRegister){
        LoginRegisterItemFragment fragment = new LoginRegisterItemFragment();
        Bundle bundle = new Bundle();
        bundle.putBoolean(BundleConstant.TYPE,isRegister);
        fragment.setArguments(bundle);
        return fragment;
    }

    private boolean isRegister;

    @Override
    public void initData() {
        isRegister = getArguments().getBoolean(BundleConstant.TYPE);
        getViewModel().isRegister().set(isRegister);
    }

    @Override
    public void initView(@NonNull View view) {

    }

    @Override
    public void createObserver() {
        getViewModel().getResponseLiveData().observe(this, new Observer<User>() {
            @Override
            public void onChanged(User user) {
                getEventVm().setUserInfo(user);
                CacheUtil.INSTANCE.setUser(user);
                QMUIFragmentExtKt.showSuccessTipsExitDialog(LoginRegisterItemFragment.this,isRegister?"注册成功":"登录成功");
            }
        });

        getViewModel().getErrorLiveData().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                showFailTipsDialog(s);
            }
        });
    }

    private void showFailTipsDialog(String msg){
        QMUIFragmentExtKt.showFailTipsDialog(LoginRegisterItemFragment.this,msg,1000,null);
    }

    @Nullable
    @Override
    public Object getBindingClick() {
        return new ProxyClick();
    }

    public class ProxyClick{

        public void login(){

            String password = getViewModel().getPassword().get();
            String rePassword = getViewModel().getRepassword().get();

            if(getViewModel().getAccount().get().isEmpty()){
                showFailTipsDialog("请填写账号");
            }else if(password.isEmpty()){
                showFailTipsDialog("请填写密码");
            }else if(isRegister&&rePassword.isEmpty()){
                showFailTipsDialog("请填写确认密码");
            }else if(isRegister&&!password.equals(rePassword)){
                showFailTipsDialog("两次密码输入不一致");
            }else{
                getViewModel().requestServer();
            }

        }

    }

}

File: app/src/main/java/com/theone/music/ui/fragment/user/LoginRegisterItemFragment.kt
package com.theone.music.ui.fragment.user

import android.view.View
import androidx.lifecycle.Observer
import androidx.work.*
import com.theone.common.constant.BundleConstant.TYPE
import com.theone.common.ext.bundle
import com.theone.common.ext.getColor
import com.theone.common.ext.getValueNonNull
import com.theone.music.R
import com.theone.music.app.util.CacheUtil
import com.theone.music.data.model.User
import com.theone.music.databinding.FragmentLoginRegister2Binding
import com.theone.music.domain.work.LoginSignWorker
import com.theone.music.viewmodel.EventViewModel
import com.theone.music.viewmodel.LoginRegisterViewModel2
import com.theone.mvvm.ext.getAppViewModel
import com.theone.mvvm.core.base.fragment.BaseCoreFragment
import com.theone.mvvm.ext.qmui.showFailTipsDialog
import com.theone.mvvm.ext.qmui.showSuccessTipsExitDialog
import java.util.concurrent.TimeUnit

class LoginRegisterItemFragment private constructor() :
    BaseCoreFragment<LoginRegisterViewModel2, FragmentLoginRegister2Binding>() {

    companion object {
        fun newInstant(isRegister: Boolean): LoginRegisterItemFragment {
            return LoginRegisterItemFragment().bundle {
                putBoolean(TYPE, isRegister)
            }
        }
    }

    private val mAppVm: EventViewModel by lazy { getAppViewModel<EventViewModel>() }

    private val isRegister: Boolean by getValueNonNull(TYPE)

    override fun getBindingClick(): Any = ProxyClick()

    override fun initView(root: View) {
        root.setBackgroundColor(getColor(mActivity, R.color.white))
        getViewModel().isRegister.set(isRegister)
    }

    override fun createObserver() {
        getViewModel().run {
            getResponseLiveData().observe(this@LoginRegisterItemFragment) {
                mAppVm.setUserInfo(it)
                CacheUtil.setUser(it)
                it.doWorker()
                showSuccessTipsExitDialog(if (isRegister.get()) "注册成功" else "登录成功")
            }
            getErrorLiveData().observe(this@LoginRegisterItemFragment) {
                showFailTipsDialog(it)
            }
        }
    }

    /**
     * 开启每日登录签到任务
     * @receiver User
     */
    private fun User.doWorker() {
        // 添加约束，网络连接时才启用
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)  // 网络状态
            .setRequiresBatteryNotLow(false)                 // 不在电量不足时执行
            .setRequiresCharging(false)                      // 在充电时执行
            .setRequiresStorageNotLow(false)                 // 不在存储容量不足时执行
            //.setRequiresDeviceIdle(true)                    // 在待机状态下执行，需要 API 23
            .build()

        val inputData = Data.Builder().putString(LoginSignWorker.EMAIL, account)
            .putString(LoginSignWorker.PASSWORD, password).build()

        // 周期任务  一天执行一次签到
        val signRequest = PeriodicWorkRequest
            .Builder(LoginSignWorker::class.java, 1, TimeUnit.DAYS)
            .setConstraints(constraints)
            .setInputData(inputData)
            .addTag(LoginSignWorker.TAG)
            .build()

        WorkManager.getInstance(mActivity).enqueue(signRequest)
    }

    inner class ProxyClick {

        fun login() {
            with(getViewModel()) {
                when {
                    isRegister.get() && email.get().isEmpty() -> showFailTipsDialog("请填写邮箱")
                    username.get().isEmpty() -> showFailTipsDialog("请填写账号")
                    password.get().isEmpty() -> showFailTipsDialog("请填写密码")
                    isRegister.get() && repassword.get().isEmpty() -> showFailTipsDialog("请填写确认密码")
                    else -> {
                        requestServer()
                    }
                }
            }
        }
    }

}